lessons/homework/ostrava5_TDD/index.md

```diff
@@ -0,0 +1,90 @@
+
+Toto je bonusový materiál, který může udělat domácí úkol zajímavější a naučit tě při tom opět něco nového. Pokud máš ale plné ruce práce se zvládání aktuální látky. klidně postupuj jen podle zadání domácích projektů a na testování se společně podíváme později.
+
+## Jak na to? ##
+
+1. Stáhni si připravený soubor pro [piškvorky][piskvorky] a [soubor s testy][testpiskvorky] a ulož je do samostatné složky.
+
+    V souboru *test_piskvorky.py* máš připravené testy. Nic v něm neměň, ale můžeš se do něj podívat. Samotnou hru pak piš do připraveného souboru _piskvorky.py_. Máš tam nachystané všechny potřebné funkce, jen zatím nic nedělají. Až je všechny doplníš, budeš mít funkční hru.
+
+1. Aktivuj si své virtuální prostředí. Jak na to jsme si ukázali [na začátku kurzu][venvsetup].
+1. Nainstaluj si knihovnu [pytest] podle [návodu][testing] v materiálech. Právě ta ti umožní ověřit správnou funkčnost programu.
+
+    ```shell
+    (venv) $ pip install pytest
+    ```
+
+1. Opět podle návodu pytest spusť.
+
+    ```shell
+    (venv) $ pytest -v test_piskvorky.py
+    ```
+
+## Co s tím? ##
+
+Pytest ti vypíše, kde všude v tvém programu narazil na problém. Tyto jsou podrobně rozepsané a výstup je tak trochu dlouhý. Na jeho začátku však budeš mít takovéto řádky:
+
+```
+test_piskvorky.py::test_vyhodnot_vyhra_x FAILED
+test_piskvorky.py::test_vyhodnot_vyhra_o FAILED
+test_piskvorky.py::test_vyhodnot_remiza FAILED
+…
+test_piskvorky.py::test_tah_pocitace_skoro_plne_konec_2 FAILED
+```
+
+Každé _FAILED_ znamená jednu chybu: jeden test, který neprošel. Protože jsi zatím nic nenapsala, je problém úplně ve všem. To se ale změní, jakmile splníš první úkol: zařídíš, aby funkce _vyhodnot_ rozpoznala, že vyhrál hráč s křížky.
+
+```
+test_piskvorky.py::test_vyhodnot_vyhra_o FAILED
+test_piskvorky.py::test_vyhodnot_vyhra_x PASSED
+test_piskvorky.py::test_vyhodnot_remiza FAILED
+…
+test_piskvorky.py::test_tah_pocitace_skoro_plne_konec_2 FAILED
+```
+
+Vidíš? Výsledek prvního testu se změnil na _PASSED_. To znamená, že v tomto případě program funguje, jak má.
+
+Až budeš mít program hotový, místo všech červených _FAILED_ bude u všech testů zelené _PASSED_. Pak víš, že máš hotovo. Teda, skoro.
+
+```
+test_piskvorky.py::test_vyhodnot_vyhra_x PASSED
+test_piskvorky.py::test_vyhodnot_vyhra_o PASSED
+…
+test_piskvorky.py::test_tah_pocitace_skoro_plne_konec PASSED
+test_piskvorky.py::test_tah_pocitace_skoro_plne_konec_2 PASSED
+```
+
+Proč jen skoro? Některé věci testovat moc dobře nejdou, nebo by to bylo pro tebe v tuto chvíli moc složité.
+
+* Jednou takovou věcí je zadání od uživatele. Funkci _tah_hrace_ proto pytest netestuje. To pak platí i pro celé jádro hry, funkci _piskvorky1d_, která právě funkci _tah_hrace_ používá, když je hráč na tahu.
```

Comment:
```suggestion
* Jednou takovou věcí je zadání od uživatele. Funkci *tah_hrace* proto pytest netestuje. To pak platí i pro celé jádro hry, funkci _piskvorky1d_, která právě funkci *tah_hrace* používá, když je hráč na tahu.
```

Tak se to opravdu rozbíjí ta podtržítka. 🙁 Na GitHubu se nerozbíjejí. Xindl toto.